### DISCUSSION TITLE
Bloom Shader setup example

### DISCUSSION THREAD

**Discussion by Psi25** [bnpr/Malt#315] [Q&A] [answered]
Does anyone have a node setup example for a bloom effect? I would like to study it.

**Comment by pragma37** [comment, marked as answer]
![imagen](https://user-images.githubusercontent.com/36610999/172175969-cf51fd3e-cd66-42e3-aaeb-5cc735cb1d0e.png)
[bloom-sample.zip](https://github.com/bnpr/Malt/files/8844966/bloom-sample.zip)

**Reply by Psi25** [reply]
I didn't know there's a screen pass. It's realtime compositing. Hella cool.

**Reply by Kolupsy** [reply]
> ![imagen](https://user-images.githubusercontent.com/36610999/172175969-cf51fd3e-cd66-42e3-aaeb-5cc735cb1d0e.png) [bloom-sample.zip](https://github.com/bnpr/Malt/files/8844966/bloom-sample.zip)

@pragma37 
Huh that is very cool. I guess I never tried to use the jitter blur for bloom. Always tried to use the other ones at low radii (because otherwise they would just kill render times) but maybe there is still a way to get bloom without fireflies. Can we sample a lower-resolution version of the screen so that lower radii can be used while still having the needed screen coverage?

**Reply by pragma37** [reply]
> I guess I never tried to use the jitter blur for bloom.

As long as you can accumulate the result over multiple samples, jitter blur is the fastest option.

> Can we sample a lower-resolution version of the screen so that lower radii can be used while still having the needed screen coverage?
Not at the moment, but it may make sense to provide a downsample render node?

I think the best option would be to provide a 2 pass blur render node, though.

You can also apply the jitter blur in one pass and then apply a small box blur on the next pass to remove the noise.
[bloom-sample-2-steps.zip](https://github.com/bnpr/Malt/files/8862346/bloom-sample-2-steps.zip)

**Reply by Kolupsy** [reply]
> You can also apply the jitter blur in one pass and then apply a small box blur on the next pass to remove the noise.

I tried that but wasnt too happy with the result to be honest. The jitter still tends to reveal itself

> I think the best option would be to provide a 2 pass blur render node, though.

Does that mean its not possible to sample a lower-resolution version of the input texture?

In general the results of the jitter blur where really nice and looked a lot like the eevee blur except of course for the fireflies

**Reply by pragma37** [reply]
> I tried that but wasnt too happy with the result to be honest. The jitter still tends to reveal itself

Have you looked at the 2 steps example? 
![imagen](https://user-images.githubusercontent.com/36610999/172642298-36edf952-2fb9-4ebb-bea5-13c1a954cd9e.png)

> Does that mean its not possible to sample a lower-resolution version of the input texture?

It's possible as long as the mips are generated beforehand, but it would be very wasteful to generate mips for every screen render target just in case a shader needs it.

**Reply by Kolupsy** [reply]
hm I am currently looking into creating a custom render node (like I did with custom shader nodes), maybe I can generate mips for just the texture that goes into my node. Maybe I would even be able to do multiple passes in one custom node? I havent really figured out yet what functions and parameters I need to override to make a custom render node

**Reply by pragma37** [reply]
Sure, you can do whatever inside render nodes.
There's not any difference between built-in and custom nodes.

You can take a look at the built-in nodes for reference:
https://github.com/bnpr/Malt/tree/Development/Malt/Nodes
https://github.com/bnpr/Malt/tree/Development/Malt/Pipelines/NPR_Pipeline/Nodes

The LineRender node is a good example for how to set up a screen pass node.

**Comment by Kolupsy** [comment]
So I took a look at the examples and they all subclass the `PipelineNode` class. and modules set a global `NODE` variable to the subclasses. I figured that Malt will generate nodes from these classes but how do I work with them actually? I cant to anything with them right? Do I pass that somewhere in the MaltPlugin routine? Basically how do I get from `PipelineNode` subclass to a usable MaltNode when I work with `PipelinePlugin`s?

**Reply by pragma37** [reply]
Here's a render node plugin example:
https://github.com/bnpr/Malt/tree/Development/plugins/Experimental

Also, for cases like this is best if you open a new thread since this is not really related to the original question.
